internal/provider/auth_kerberos.go

```diff
@@ -0,0 +1,232 @@
+			},
+			consts.FieldService: {
+				Type:          schema.TypeString,
+				Optional:      true,
+				Description:   "The service principle name.",
+				ConflictsWith: conflicts,
+			},
+			consts.FieldRealm: {
+				Type:          schema.TypeString,
+				Optional:      true,
+				Description:   "The Kerberos server's authoritative authentication domain",
+				ConflictsWith: conflicts,
+			},
+			consts.FieldKRB5ConfPath: {
+				Type:          schema.TypeString,
+				Optional:      true,
+				Description:   "A valid Kerberos configuration file e.g. /etc/krb5.conf.",
+				DefaultFunc:   schema.EnvDefaultFunc(consts.EnvVarKRB5Conf, nil),
+				ValidateFunc:  validateFileExists,
+				ConflictsWith: conflicts,
+			},
+			consts.FieldKeytabPath: {
+				Type:          schema.TypeString,
+				Optional:      true,
+				Description:   "The Kerberos keytab file containing the entry of the login entity.",
+				DefaultFunc:   schema.EnvDefaultFunc(consts.EnvVarKRBKeytab, nil),
+				ValidateFunc:  validateFileExists,
+				ConflictsWith: conflicts,
+			},
+			consts.FieldDisableFastNegotiation: {
+				Type:          schema.TypeBool,
+				Optional:      true,
+				Default:       false,
+				ConflictsWith: conflicts,
+				Description:   "Disable the Kerberos FAST negotiation.",
+			},
+			consts.FieldRemoveInstanceName: {
+				Type:          schema.TypeBool,
+				Optional:      true,
+				Default:       false,
+				ConflictsWith: conflicts,
+				Description:   "Strip the host from the username found in the keytab.",
+			},
+		},
+	}, consts.MountTypeKerberos)
+
+	return s
+}
+
+type AuthLoginKerberos struct {
+	AuthLoginCommon
+	// useful for unit testing
+	authHeaderFunc func(*krbauth.LoginCfg) (string, error)
+}
+
+// MountPath for the kerberos authentication engine.
+func (l *AuthLoginKerberos) MountPath() string {
+	if l.mount == "" {
+		return l.Method()
+	}
+	return l.mount
+}
+
+// LoginPath for the kerberos authentication engine.
+func (l *AuthLoginKerberos) LoginPath() string {
+	return fmt.Sprintf("auth/%s/login", l.MountPath())
+}
+
+func (l *AuthLoginKerberos) Init(d *schema.ResourceData, authField string) error {
+	if err := l.AuthLoginCommon.Init(d, authField); err != nil {
+		return err
+	}
+
+	if _, ok := l.getOk(d, consts.FieldToken); !ok {
+		required := []string{
+			consts.FieldUsername,
+			consts.FieldService,
+			consts.FieldRealm,
+			consts.FieldKeytabPath,
+			consts.FieldKRB5ConfPath,
+		}
+		var missing []string
+		for _, f := range required {
+			if _, ok := l.getOk(d, f); !ok {
+				missing = append(missing, f)
+			}
+		}
+
+		if len(missing) > 0 {
+			return fmt.Errorf("required fields are unset: %v", missing)
+		}
+	}
+
+	return nil
+}
+
+// Method name for the kerberos authentication engine.
+func (l *AuthLoginKerberos) Method() string {
+	return consts.AuthMethodKerberos
+}
+
+// Login using the kerberos authentication engine.
+func (l *AuthLoginKerberos) Login(client *api.Client) (*api.Secret, error) {
+	if !l.initialized {
```

Comment:
Is this something we want to check for other auth methods as well? From first glance it would seem like we always want to ensure params are populated before we attempt to login

Reply:
Defintely, I was planning to add it later in another PR.